### PR TITLE
[Merged by Bors] - sql: ref ballots

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation

Part of https://github.com/spacemeshos/go-spacemesh/issues/2918

## Changes

- move ref ballots storage to SQLite

## Test Plan

Unit and system tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by nkryuchkov** [comment]
bors try

File: sql/migrations/0001_initial.sql [hunk 1]
```diff
@@ -17,6 +17,12 @@ CREATE TABLE ballots
 );
 CREATE INDEX ballots_by_layer_by_pubkey ON ballots (layer, pubkey);
 
+CREATE TABLE ref_ballots
```
Comment (countvonzero) [inline sql/migrations/0001_initial.sql]: why does ref ballot need to be a separate table and not a field of ballots table?

btw, the primary key should be (epoch, smesher public key)

**Comment by dshulyak** [comment]
@nkryuchkov this change is not crash safe. 

1. save ref ballot
2. crash
3. save ballot

when app recovers we will have a pointer, but not the ballot itself. note that we can't simply reverse the order, as it will result in a different problem.

as i mentioned this approach is ok only if pointer is saved atomically with the ballot itself. after reviewing code we can actually reuse existing index to find ref ballot (`ballots_by_layer_by_pubkey`) . the ref ballot would be the first ballot in the layer from epoch range. i think we should use that approach since it is both crash-safe and simplifies db access (we actually need only one additional read)

**Comment by nkryuchkov** [comment]
bors try

File: sql/ballots/ballots.go [hunk 2]
```diff
@@ -148,3 +148,28 @@ func CountByPubkeyLayer(db sql.Executor, lid types.LayerID, pubkey []byte) (int,
...
 	return rows, nil
 }
+
+// GetRefBallot gets a ref ballot for a layer and a pubkey.
+func GetRefBallot(db sql.Executor, epochID types.EpochID, pubkey []byte) (ballotID types.BallotID, err error) {
+	firstLayer := epochID.FirstLayer()
+	lastLayer := firstLayer.Add(types.GetLayersPerEpoch())
+	rows, err := db.Exec(`
+		select * from ballots 
+		where layer between ?1 and ?2 and pubkey = ?3
```
Comment (dshulyak) [inline sql/ballots/ballots.go:158]: `between` keyword inclusive or exclusive? would be nice to have a test for this case as well

**Comment by dshulyak** [comment]
please add a test for new function, also i don't remember if index is sorted in asc order by default, it is better to make this change 

```
-CREATE INDEX ballots_by_layer_by_pubkey ON ballots (layer, pubkey);
+CREATE INDEX ballots_by_layer_by_pubkey ON ballots (layer asc, pubkey);
```

File: miner/proposal_builder_test.go [hunk 3]
```diff
@@ -297,55 +282,13 @@ func TestBuilder_HandleLayer_BaseBlockError(t *testing.T) {
 	assert.ErrorIs(t, b.handleLayer(context.TODO(), layerID), errUnknown)
 }
 
-func TestBuilder_HandleLayer_GetRefBallotError(t *testing.T) {
```
Comment (countvonzero) [inline miner/proposal_builder_test.go:300]: can you change this test to use sqlite?

File: sql/ballots/ballots.go [hunk 4]
```diff
@@ -148,3 +148,28 @@ func CountByPubkeyLayer(db sql.Executor, lid types.LayerID, pubkey []byte) (int,
 	}
 	return rows, nil
 }
+
+// GetRefBallot gets a ref ballot for a layer and a pubkey.
+func GetRefBallot(db sql.Executor, epochID types.EpochID, pubkey []byte) (ballotID types.BallotID, err error) {
+	firstLayer := epochID.FirstLayer()
+	lastLayer := firstLayer.Add(types.GetLayersPerEpoch()).Sub(1)
+	rows, err := db.Exec(`
+		select * from ballots
```
Comment (countvonzero) [inline sql/ballots/ballots.go]: in code you are consistent about getting ballot including the malicious field. i think you should do it here too.
i think the query should be the same as `Get()` with the extra where clause

File: sql/ballots/ballots.go [hunk 5]
```diff
@@ -148,3 +148,28 @@ func CountByPubkeyLayer(db sql.Executor, lid types.LayerID, pubkey []byte) (int,
...
+			stmt.BindInt64(2, int64(lastLayer.Value))
+			stmt.BindBytes(3, pubkey)
+		}, func(stmt *sql.Statement) bool {
+			stmt.ColumnBytes(0, ballotID[:])
+			return true
+		})
+	if err != nil {
+		return types.BallotID{}, fmt.Errorf("ref ballot epoch %v: %w", epochID, err)
+	} else if rows == 0 {
+		return types.BallotID{}, fmt.Errorf("%w ballot epoch %s", sql.ErrNotFound, epochID)
```
Comment (countvonzero) [inline sql/ballots/ballots.go]: ref ballot epoch

File: miner/proposal_builder_test.go (hunk 3 above)
Comment (countvonzero) [reply miner/proposal_builder_test.go:300]: done

File: sql/ballots/ballots.go (hunk 4 above)
Comment (countvonzero) [reply sql/ballots/ballots.go]: changed to `select id` instead

File: sql/ballots/ballots.go (hunk 5 above)
Comment (countvonzero) [reply sql/ballots/ballots.go]: done

**Review by countvonzero — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by nkryuchkov** [comment]
@countvonzero thanks for fixing

**Pull request closed without merging**